dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -296,37 +302,102 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
     }
 
+#if 1 //  Option 1
     winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification>> AppNotificationManager::GetAllAsync()
     {
-        co_await winrt::resume_background();
+        HRESULT telemetryResult{ S_OK };
+        std::exception_ptr eptr{};
+        std::wstring appId{};
 
-        std::wstring appId{ RetrieveNotificationAppId() };
-        winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection{};
-        auto result{ ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()) };
+        auto logTelemetry{ wil::scope_exit([&](){
+            AppNotificationTelemetry::GetAllAsyncByAPI(telemetryResult, appId);
+            rethrowIf(eptr);
+        }) };
+
+        try
+        {
+            co_await winrt::resume_background();
+
+            appId = RetrieveNotificationAppId();
+
+            winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection{};
+            auto result{ ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()) };
+
+            THROW_HR_IF(result, result != S_OK && result != E_NOT_SET); // Swallow E_NOT_SET and return an empty properties vector to signal that there are no active toasts
+
+            unsigned int count{};
+            if (toastPropertiesCollection)
+            {
+                THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            }
+
+            winrt::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification> toastNotifications{ winrt::single_threaded_vector<winrt::Microsoft::Windows::AppNotifications::AppNotification>() };
+
+            for (unsigned i = 0; i < count; ++i)
+            {
+                ToastABI::INotificationProperties* toastProperties;
+                THROW_IF_FAILED(toastPropertiesCollection->GetAt(i, &toastProperties));
 
-        THROW_HR_IF(result, result != S_OK && result != E_NOT_SET); // Swallow E_NOT_SET and return an empty properties vector to signal that there are no active toasts
+                auto toastNotification{ ToastNotificationFromToastProperties(toastProperties) };
 
-        unsigned int count{};
-        if (toastPropertiesCollection)
+                toastNotifications.Append(toastNotification);
+            }
+
+            co_return toastNotifications;
+        }
+        catch (...)
         {
-            THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            eptr = std::current_exception();
+            telemetryResult = wil::ResultFromCaughtException();
         }
+    }
+#else // Option 2
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification>> AppNotificationManager::GetAllAsync()
```

Comment:
Second option, which handles successes and failures in separate path. Mostly offered for reference and comparison, I'm slowly warming up to the first option.

Reply:
I agree with the second option as well 😊.

Reply:
Are you waiting for more thoughts on the approach?

Reply:
Take a look at my comments above. I think that simplifies option 1.

Reply:
We're going with the first option.